subsys/modbus/modbus_server.c

```diff
@@ -345,6 +345,9 @@ static bool mbs_fc03_hreg_read(struct modbus_context *ctx)
 			return true;
 		}
 
+		/* compute the register quantity in terms of 16-bit registers */
+		reg_qty = reg_qty / 2;
```

Comment:
A few more changes are needed to fix this, the server side should also pass the correct address to the user callback. And we also need to fix the client side. Please have a look at https://github.com/zephyrproject-rtos/zephyr/pull/78556

Reply:
Hi, I agree with your point, and I've tested server side of the #78556 - which addresses the problem and is a better solution. I will close this PR.